### ISSUE TITLE
On startup if Umbraco cannot connect to the database, it always opens the installer

### ISSUE THREAD

**Issue by joepvtl** [umbraco/Umbraco-CMS#12949]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

10.1.0

### Bug summary

When Umbraco can't connect to the database, the installer page opens even if the Umbraco environment was already configured.

### Specifics

Using the Provider: Microsoft.Data.SqlClient

```
[11:27:08 DBG] Could not immediately connect to database, trying again.
[11:27:09 WRN] Configured database is reporting as not being available.
Microsoft.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: Named Pipes Provider, error: 40 - Could not open a connection to SQL Server)
 ---> System.ComponentModel.Win32Exception (53): The network path was not found.
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
   at Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at Microsoft.Data.ProviderBase.DbConnectionClosed.TryOpenConnection(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at Microsoft.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry, SqlConnectionOverrides overrides)
   at Microsoft.Data.SqlClient.SqlConnection.Open(SqlConnectionOverrides overrides)
   at Microsoft.Data.SqlClient.SqlConnection.Open()
   at Umbraco.Extensions.DbConnectionExtensions.IsAvailable(IDbConnection connection)
ClientConnectionId:00000000-0000-0000-0000-000000000000
Error Number:53,State:0,Class:20
[11:27:09 DBG] Could not connect to database.
[11:27:09 DBG] Runtime level: Install - InstallMissingDatabase
```

I think this bug got introduced in this commit: https://github.com/umbraco/Umbraco-CMS/commit/8e6e262c7f0e3985014dd29cd635f2ac7a286b0f#diff-6ac9a3ad41721ad6c1743713a759eef7e43bd731f3ce6294023d2af3b7e7dd0eL146

### Steps to reproduce

1. Setup a Umbraco environment with SQL server that doesn't run on your localdb.
2. Now change the server part in the connectionstring to a domain that doesn't exist for example: test.test
3. Restart your Umbraco environment and the install page should show.

### Expected result / actual result

**Expected**
I expect to see a 500 page with the message: "A connection string is configured but Umbraco could not connect to the database." like in Umbraco 9.

**Actual**
Actual result is a installer page that everyone can see.

---
_This item has been added to our backlog [AB#22604](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/22604)_

**Comment by nul800sebastiaan** [comment]
I haven't tried this yet, does sound like a bug, but I am a bit curious as to what the use case is here? I understand your example is made for us to be able to reproduce it, but what are you actually trying to accomplish by changing the connection string?

**Comment by joepvtl** [comment]
Its just an example to reproduce for when Umbraco can't reach your database. In earlier versions Umbraco would just show a 500 page, but now it opens up the installer which everyone can see and configure. I feel like this should never just happen when the database can't be reached.

**Comment by nul800sebastiaan** [comment]
Yes, that is a good point, shouldn't jump to the installer if the database can't be reached, didn't think of that, thanks! We'll have a look.

**Comment by nul800sebastiaan** [comment]
I can reproduce the behavior in v10.0.1 and also in v9.0.1 as well, so it hasn't changed in the referred commit above.

**Comment by joepvtl** [comment]
When I tested this in v9 It does have the expected behaviour in v9.5.4

```
[10:48:24 DBG] Could not immediately connect to database, trying again.
[10:48:25 WRN] Configured database is reporting as not being available.
System.Data.SqlClient.SqlException (0x80131904): A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: Named Pipes Provider, error: 40 - Could not open a connection to SQL Server)
 ---> System.ComponentModel.Win32Exception (53): The network path was not found.
   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
   at System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at System.Data.ProviderBase.DbConnectionClosed.TryOpenConnection(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at System.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry)
   at System.Data.SqlClient.SqlConnection.Open()
   at Umbraco.Extensions.DbConnectionExtensions.IsAvailable(IDbConnection connection)
ClientConnectionId:00000000-0000-0000-0000-000000000000
Error Number:53,State:0,Class:20
[10:48:25 DBG] Could not connect to database.
[10:48:25 ERR] Boot Failed
Umbraco.Cms.Core.Exceptions.BootFailedException: A connection string is configured but Umbraco could not connect to the database.
   at Umbraco.Cms.Infrastructure.Runtime.RuntimeState.DetermineRuntimeLevel()
   at Umbraco.Cms.Infrastructure.Runtime.CoreRuntime.DetermineRuntimeLevel()

```

**Comment by nul800sebastiaan** [comment]
Hmmm, this is what I'm seeing:


https://user-images.githubusercontent.com/304656/188593600-828053a3-df61-4620-a7df-9e206cb4ca4b.mp4


But maybe this is specifically about SQL Server, I didn't test that.

**Comment by joepvtl** [comment]
Yeah this scenario does exist in v9 when the connectionstring has the value localdb in it (see [this line](https://github.com/umbraco/Umbraco-CMS/commit/8e6e262c7f0e3985014dd29cd635f2ac7a286b0f#diff-6ac9a3ad41721ad6c1743713a759eef7e43bd731f3ce6294023d2af3b7e7dd0eL316)) But I guess that was intended behaviour because in the commit I referred to, the old code did check for that value and if it has that value, it sets the state to InstallMissingDatabase. But now it always does this.

**Comment by nul800sebastiaan** [comment]
So I just tried the exact scenario from the video above in 9.5.4, but I used SQL Server this time. I stopped the site, changed the server from `localhost` to `test.test` and started the site again, it launches into the Umbraco installer screen, like above.

**Comment by nul800sebastiaan** [comment]
In 10.1.1 when I stop the server so that it can't be reached any more, while the site is running, I get the expected error, and that is also what you were expecting to see 

> when Umbraco can't reach your database, when your database server is offline or has a a hiccup. In earlier versions Umbraco would just show a 500 page

![image](https://user-images.githubusercontent.com/304656/188599535-b791d863-963c-4631-a6e9-1c642adbe801.png)

**Comment by joepvtl** [comment]
Yeah the issue is only on boot, not when the site is already running.

**Comment by rammi987** [comment]
@nul800sebastiaan Meaby this can help.

When you have an Connectionstring, witch are not null or empty, we end here
[https://github.com/umbraco/Umbraco-CMS/blob/v10/contrib/src/Umbraco.Infrastructure/Runtime/RuntimeState.cs#L213](url)

as for the Global setting `InstallMissingDatabase` is per default false in 9 and 10, but if we take a closer look into `_databaseProviderMetadata.CanForceCreateDatabase(_databaseFactory.ProviderName)` there is the iconicity happing.

[https://github.com/umbraco/Umbraco-CMS/blob/v10/contrib/src/Umbraco.Cms.Persistence.SqlServer/Services/SqlLocalDbDatabaseProviderMetadata.cs#L52](url)
[https://github.com/umbraco/Umbraco-CMS/blob/v10/contrib/src/Umbraco.Cms.Persistence.SqlServer/Services/SqlAzureDatabaseProviderMetadata.cs#L51](url)
[https://github.com/umbraco/Umbraco-CMS/blob/v10/contrib/src/Umbraco.Cms.Persistence.SqlServer/Services/SqlServerDatabaseProviderMetadata.cs#L50](url)
[https://github.com/umbraco/Umbraco-CMS/blob/v10/contrib/src/Umbraco.Cms.Persistence.Sqlite/Services/SqliteDatabaseProviderMetadata.cs#L46](url)

I testede with the Sqlite where I changes the database name to get an invalid connection string, but ended on the install page. `_databaseProviderMetadata.CanForceCreateDatabase(_databaseFactory.ProviderName)`  is returning true, due to hardcoded force at [https://github.com/umbraco/Umbraco-CMS/blob/v10/contrib/src/Umbraco.Cms.Persistence.Sqlite/Services/SqliteDatabaseProviderMetadata.cs#L46](url)

For Umbraco 9.
[https://github.com/umbraco/Umbraco-CMS/blob/v9/dev/src/Umbraco.Infrastructure/Runtime/RuntimeState.cs#L313](url)

In 9 and 10, if using either Sqlite or SQL Server Express LocalDB it will end up in install page.

**Comment by JoseMarcenaro** [comment]
I can confirm this is still happening in v10.1.0 and v10.2.0
It is a problem when connectivity with the SQL server is broken due to any reason - i.e. firewall issues on Azure.

**Comment by gilbertaoe** [comment]
I saw a separate open issue with this same problem and have already made the fix and created a pull request. I accidentally added this issue's link to the wrong pull request... so ignore the mention of the X-XSS-Protection pull requets :p

Pull request: https://github.com/umbraco/Umbraco-CMS/pull/13045
Duplicate Issue: https://github.com/umbraco/Umbraco-CMS/issues/13042

**Comment by gilbertaoe** [comment]
@nul800sebastiaan to clarify the issue further if it helps, for 10.2.0, I could only replicate if I stopped and started the app pool. If I didn't restart the app pool, due to the nature of Umbraco's Backoffice, restricting the database's firewall didn't immediately cause any issues.

**Comment by JoseMarcenaro** [comment]
In my case (Azure with slots) the issue happens when swapping the slots, if for some reason the DB is out of reach (i.e. misconfigured network firewall)

**Comment by gilbertaoe** [comment]
This should "fix" that problem for you... it will show you an error pge that the DB is out of reach, but it won't prompt you to re-install.

**Comment by JoseMarcenaro** [comment]
Thank @gilbertaoe .  Reaching the Install page is really bad, because once you did, you are stuck there even after you have restored the DB connectivity. 
As far as I know, you need to manually empty the `/umbraco/data` folder to get rid of the install stuff.

**Comment by kjac** [comment]
Whoops this one was fixed in #13118 🙈
